cli/azd/pkg/executil/executil.go

```diff
@@ -18,6 +18,26 @@ import (
 	"strings"
 )
 
+type RunWithResultFn func(ctx context.Context, args RunArgs) (RunResult, error)
+type contextKey string
+
+const (
+	execFnContextKey contextKey = "execfn"
+)
+
+func WithExecUtil(ctx context.Context, execFn RunWithResultFn) context.Context {
```

Comment:
During normal flow the `execFn` is not stored in context and is returned each time.  This was primarily added to support unit testing where we can mock the `execFn` that is used in most of our CLI command wrappers.

During test setup the mock `execFn` is set with `WithExecUtil(...)` and then retrieved later when commands are executed.  This allows for handling a variety of mock situations from azcli, docker cli, bicep cli, etc.

Reply:
So it would be difficult to restrict that aspect to test code only?

Reply:
Since all the CLI wrappers use the `execFn` my goal was to have this in a common place vs the dev having to mock the full `AzCli` interface (or others) that contains dozens of methods.  By supporting mocking at the exec later we enable easier mocking against and CLI wrapper.

If we really wanted to we could perform a check in `WithExecUtil(...)` method to panic if it was attempted to be set in a non-test env.

Reply:
Updated comments in code to be more clear of intended usage.

Reply:
Based on the team discussion yesterday, it sounds like we're having a design change to use `context` as an IoC with this change. It's a temporary one while we look at other options.

The current proposed design based on this PR (if I understand correctly):
1. `commands.Build` is the current forced entry point for all commands. Object graph is constructed and stored into the `ctx` here.
2. Any command would grab dependencies from `ctx`.

I'm comfortable with using `context` as a temporary solution, but only if:
1. `Get<Something>` methods need to panic by default unless something is truly optional.
2. `Get<OptionalSomething>` returns `error` if not found in `context`.

Anything that implicitly returns a default is potential concern for implicit, undefined behavior when a programming mistake occurs.